control-plane/csi-driver/src/bin/node/nodeplugin_nvme.rs

```diff
@@ -1,10 +1,68 @@
-use crate::dev::Device;
+use crate::dev::{Device, DeviceError};
 use grpc::csi_node_nvme::{
     nvme_operations_server::NvmeOperations, NvmeConnectRequest, NvmeConnectResponse,
 };
-use std::collections::HashMap;
+use once_cell::sync::OnceCell;
+use std::{
+    collections::{HashMap, HashSet},
+    sync::Mutex,
+};
 use tonic::Response;
-use tracing::{info, warn};
+use tracing::{error, info, trace, warn};
+
+static REQUEST_INVENTORY: OnceCell<Mutex<HashSet<String>>> = OnceCell::new();
+
+/// Initialise the REQUEST_INVENTORY once cell.
+pub(crate) fn init_request_inventory(data: HashSet<String>) {
+    REQUEST_INVENTORY.get_or_init(|| Mutex::new(data));
+}
+
+/// Returns the REQUEST_INVENTORY.
+pub(crate) fn get_request_inventory() -> Result<&'static Mutex<HashSet<String>>, DeviceError> {
+    REQUEST_INVENTORY
+        .get()
+        .ok_or_else(|| DeviceError::new("Could not access cache"))
+}
+
+struct RequestEntry {
+    item: String,
+}
+
+impl RequestEntry {
+    /// Creates a new RequestEntry.
+    fn new(item: String) -> Self {
+        Self { item }
+    }
+
+    /// Inserts the held item to REQUEST_INVENTORY.
+    fn insert(&self) -> Result<(), DeviceError> {
+        let mut cache = get_request_inventory()?.lock().unwrap();
+        if cache.get(self.item.as_str()).is_some() {
+            trace!(
+                self.item,
+                "Uri already exists in cache, skipping connection"
+            );
+            return Err(DeviceError::new("Nvme connect in progress"))?;
+        } else {
+            cache.insert(self.item.clone());
+            trace!(self.item, "Uri inserted in cache, starting connection")
+        }
+        Ok(())
+    }
+}
+
+impl Drop for RequestEntry {
+    fn drop(&mut self) {
+        if let Ok(inventory) = get_request_inventory() {
```

Comment:
this is still not sound, the second request will actually free the first here?
Consider using RAII instead by using a single operation as I've suggested.

Reply:
Ahh right. Understood. Would make that a single operation.

Reply:
Done.